File: 09/gemini-3.0-pro/go/main.go
package main

import (
"bufio"
"fmt"
"math"
"os"
"strconv"
"strings"
)

type Point struct {
x, y int64
}

type Edge struct {
	p1, p2 Point
}

func isPointInPolygon(x, y float64, edges []Edge) bool {
	inside := false
	for _, edge := range edges {
		p1 := edge.p1
		p2 := edge.p2
		if (float64(p1.y) > y) != (float64(p2.y) > y) {
			intersectX := float64(p2.x-p1.x)*(y-float64(p1.y))/(float64(p2.y)-float64(p1.y)) + float64(p1.x)
			if x < intersectX {
				inside = !inside
			}
		}
	}
	return inside
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func isValid(p1, p2 Point, edges []Edge) bool {
	x1 := min(p1.x, p2.x)
	x2 := max(p1.x, p2.x)
	y1 := min(p1.y, p2.y)
	y2 := max(p1.y, p2.y)

	cx := float64(x1+x2) / 2.0
	cy := float64(y1+y2) / 2.0

	if !isPointInPolygon(cx, cy, edges) {
		return false
	}

	for _, edge := range edges {
		ep1 := edge.p1
		ep2 := edge.p2

		if ep1.x == ep2.x { // Vertical
			ex := ep1.x
			eyMin := min(ep1.y, ep2.y)
			eyMax := max(ep1.y, ep2.y)
			if x1 < ex && ex < x2 {
				overlapMin := max(y1, eyMin)
				overlapMax := min(y2, eyMax)
				if overlapMin < overlapMax {
					return false
				}
			}
		} else if ep1.y == ep2.y { // Horizontal
			ey := ep1.y
			exMin := min(ep1.x, ep2.x)
			exMax := max(ep1.x, ep2.x)
			if y1 < ey && ey < y2 {
				overlapMin := max(x1, exMin)
				overlapMax := min(x2, exMax)
				if overlapMin < overlapMax {
					return false
				}
			}
		}
	}
	return true
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var points []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}
		parts := strings.Split(line, ",")
		x, _ := strconv.ParseInt(parts[0], 10, 64)
		y, _ := strconv.ParseInt(parts[1], 10, 64)
		points = append(points, Point{x, y})
	}

	var edges []Edge
	n := len(points)
	for i := 0; i < n; i++ {
		edges = append(edges, Edge{points[i], points[(i+1)%n]})
	}

	var maxAreaPart1 int64 = 0
	var maxAreaPart2 int64 = 0

	for i := 0; i < len(points); i++ {
		for j := i + 1; j < len(points); j++ {
			p1 := points[i]
			p2 := points[j]

			width := int64(math.Abs(float64(p1.x - p2.x))) + 1
			height := int64(math.Abs(float64(p1.y - p2.y))) + 1
			area := width * height

			if area > maxAreaPart1 {
				maxAreaPart1 = area
			}

			if isValid(p1, p2, edges) {
				if area > maxAreaPart2 {
					maxAreaPart2 = area
				}
			}
		}
	}

	fmt.Printf("Day 09 Part 1: %d\n", maxAreaPart1)
	fmt.Printf("Day 09 Part 2: %d\n", maxAreaPart2)
}

File: 09/gemini-3.0-pro/rust/src/main.rs
use std::fs;

#[derive(Debug, Clone, Copy)]
struct Point {
    x: i64,
    y: i64,
}

struct Edge {
    p1: Point,
    p2: Point,
}

fn is_point_in_polygon(x: f64, y: f64, edges: &[Edge]) -> bool {
    let mut inside = false;
    for edge in edges {
        let p1 = edge.p1;
        let p2 = edge.p2;
        if (p1.y as f64 > y) != (p2.y as f64 > y) {
            let intersect_x = (p2.x - p1.x) as f64 * (y - p1.y as f64) / (p2.y - p1.y) as f64 + p1.x as f64;
            if x < intersect_x {
                inside = !inside;
            }
        }
    }
    inside
}

fn is_valid(p1: Point, p2: Point, edges: &[Edge]) -> bool {
    let x1 = p1.x.min(p2.x);
    let x2 = p1.x.max(p2.x);
    let y1 = p1.y.min(p2.y);
    let y2 = p1.y.max(p2.y);

    let cx = (x1 + x2) as f64 / 2.0;
    let cy = (y1 + y2) as f64 / 2.0;

    if !is_point_in_polygon(cx, cy, edges) {
        return false;
    }

    for edge in edges {
        let ep1 = edge.p1;
        let ep2 = edge.p2;

        if ep1.x == ep2.x { // Vertical
            let ex = ep1.x;
            let ey_min = ep1.y.min(ep2.y);
            let ey_max = ep1.y.max(ep2.y);
            if x1 < ex && ex < x2 {
                let overlap_min = y1.max(ey_min);
                let overlap_max = y2.min(ey_max);
                if overlap_min < overlap_max {
                    return false;
                }
            }
        } else if ep1.y == ep2.y { // Horizontal
            let ey = ep1.y;
            let ex_min = ep1.x.min(ep2.x);
            let ex_max = ep1.x.max(ep2.x);
            if y1 < ey && ey < y2 {
                let overlap_min = x1.max(ex_min);
                let overlap_max = x2.min(ex_max);
                if overlap_min < overlap_max {
                    return false;
                }
            }
        }
    }
    true
}

fn main() {
    let content = fs::read_to_string("input.txt").expect("Failed to read input file");
    let points: Vec<Point> = content
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split(',').collect();
            Point {
                x: parts[0].parse().unwrap(),
                y: parts[1].parse().unwrap(),
            }
        })
        .collect();

    let mut edges = Vec::new();
    let n = points.len();
    for i in 0..n {
        edges.push(Edge {
            p1: points[i],
            p2: points[(i + 1) % n],
        });
    }

    let mut max_area_part1: i64 = 0;
    let mut max_area_part2: i64 = 0;

    for i in 0..points.len() {
        for j in (i + 1)..points.len() {
            let p1 = points[i];
            let p2 = points[j];

            let width = (p1.x - p2.x).abs() + 1;
            let height = (p1.y - p2.y).abs() + 1;
            let area = width * height;

            if area > max_area_part1 {
                max_area_part1 = area;
            }

            if is_valid(p1, p2, &edges) {
                if area > max_area_part2 {
                    max_area_part2 = area;
                }
            }
        }
    }

    println!("Day 09 Part 1: {}", max_area_part1);
    println!("Day 09 Part 2: {}", max_area_part2);
}
